java/ql/src/Security/CWE/CWE-798/HardcodedCredentialsApiCall.ql

```diff
@@ -26,9 +26,9 @@ class HardcodedCredentialApiCallConfiguration extends DataFlow::Configuration {
 
   override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
     node1.asExpr().getType() instanceof TypeString and
-    exists(MethodAccess ma | ma.getMethod().hasName(["getBytes", "toCharArray"]) |
+    exists(MethodAccess ma | ma.getMethod().hasName(["getBytes", "toCharArray","decode"]) |
```

Comment:
Be specific: what type's `decode` method are you referring to here? (the existing `getBytes` and `toCharArray` should be specific too in fairness, though no need to fiddle with that in this PR)

Reply:
when i test `cookieRememberMeManager.setCipherKey(Base64.decode("1234"))` by the query without `decode` , it do not work. so i add 'decode' method as additional data flow. Other queries used to detect hard-coded may also encounter similar situations.

Reply:
OK, in that case you should check if `ma.getMethod().hasQualifiedName("java.util", "Base64$Decoder", "decode")`

The existing `getBytes` and `toCharArray` methods should also be qualified that way; it isn't mandatory for you to fix that here though.

Reply:
Shouldn't this work out-of-the-box?
https://github.com/github/codeql/blob/9b0e3b19502c895730a35665e5652f3c60debdcd/java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll#L244-L251

Reply:
Good spot, yes it should. Can you show a test case that requires this extra `decode` method to be modelled this way?

Reply:
nice. but in some case, `java.util.Base64$Decoder` is not be used. for instance, `org.apache.shiro.codec.Base64` in `https://github.com/yinna/wjproject/blob/master/wj/src/main/java/com/example/wj/config/ShiroConfiguration.java`,`cn.hutool.core.codec.Base64` in `https://github.com/ph-sir/usermg/blob/master/src/main/java/com/pengh/usermg/shiro/ShiroConfig.java`, how to detect that?

Reply:
Either we would welcome a PR to model Shiro and/or Hutool, providing model specifications similar to those that @intrigus-lgtm linked to, or if you just want to provide enough support for this query, write something like

```
ma.getMethod().hasQualifiedName(["org.apache.shiro.codec", "cn.hutool.core.codec"], "Base64", "decode")
```

The important thing is to specify particular methods you want to model, rather than guessing that any future method encountered that is named `decode` will behave the same way.